### PR TITLE
feat(grpc): enforce a maximum number of reconnection attempts

### PR DESCRIPTION
**Pull request by Molter73**
## Description

If fact fails to connect to the gRPC server in the specified number of attemtps it will crash, which can be used as a clear signal in k8s that something is not working as expected. If the number of retries is set to 0, there will be no limit to the amount of times fact attempts to reconnect.

In order to allow the reconnection failure to trigger an application wide crash, the main output task monitors the result of the grpc task's handle propagating the error further up. This method should allow for other output components to be added in the future and follow this same pattern without having to change anything outside the output module. The stdout component is not included in this logic because it has no condition that could merit an application wide crash.

## Checklist
- [x] Patch has a change log entry **OR** does not need one.
- [x] Investigated and inspected CI test results
- [ ] Updated documentation accordingly

**Automated testing**
  - [x] Added unit tests
  - [ ] Added integration tests
  - [ ] Added regression tests

If any of these don't apply, please comment below.

## Testing Performed

Pointed fact to connect to a non-existant gRPC server and got it to crash as expected.

<details>
  <summary>fact output on crash</summary>

```
[INFO  2026-06-24T16:55:17Z] FactConfig {
    paths: Some(
        [
            "/etc/sensitive-files/**/*",
            "/etc/sensitive-files",
        ],
    ),
    grpc: GrpcConfig {
        url: Some(
            "http://localhost:9999",
        ),
        certs: None,
        backoff: BackoffConfig {
            initial: None,
            max: Some(
                1s,
            ),
            jitter: Some(
                false,
            ),
            multiplier: None,
            retries_max: Some(
                3,
            ),
        },
    },
    endpoint: EndpointConfig {
        address: None,
        expose_metrics: Some(
            true,
        ),
        health_check: None,
    },
    bpf: BpfConfig {
        ringbuf_size: None,
        inodes_max: None,
    },
    skip_pre_flight: None,
    json: None,
    hotreload: None,
    scan_interval: None,
    rate_limit: None,
}
[INFO  2026-06-24T16:55:17Z] fact version: 0.3.x-104-g669f5be002
[INFO  2026-06-24T16:55:17Z] OS: Fedora Linux 44 (Workstation Edition)
[INFO  2026-06-24T16:55:17Z] Kernel version: 7.0.12-201.fc44.x86_64
[INFO  2026-06-24T16:55:17Z] Architecture: x86_64
[INFO  2026-06-24T16:55:17Z] Hostname: mmoltras-thinkpadp1gen5.remote.csb
[INFO  2026-06-24T16:55:17Z] Starting BPF worker...
[WARN  2026-06-24T16:55:17Z] Invalid initial value: 1 >= 1
[INFO  2026-06-24T16:55:17Z] Attempting to connect to gRPC server...
[INFO  2026-06-24T16:55:17Z] Starting host scanner...
[WARN  2026-06-24T16:55:17Z] Using unencrypted gRPC channel
[INFO  2026-06-24T16:55:17Z] Failed to connect to server: transport error

Caused by:
    0: tcp connect error
    1: tcp connect error
    2: Connection refused (os error 111), retrying in 1s
[INFO  2026-06-24T16:55:18Z] Attempting to connect to gRPC server...
[WARN  2026-06-24T16:55:18Z] Using unencrypted gRPC channel
[INFO  2026-06-24T16:55:18Z] Failed to connect to server: transport error

Caused by:
    0: tcp connect error
    1: tcp connect error
    2: Connection refused (os error 111), retrying in 1s
[INFO  2026-06-24T16:55:19Z] Attempting to connect to gRPC server...
[WARN  2026-06-24T16:55:19Z] Using unencrypted gRPC channel
[INFO  2026-06-24T16:55:19Z] Failed to connect to server: transport error

Caused by:
    0: tcp connect error
    1: tcp connect error
    2: Connection refused (os error 111), retrying in 1s
[INFO  2026-06-24T16:55:20Z] Attempting to connect to gRPC server...
[WARN  2026-06-24T16:55:20Z] Using unencrypted gRPC channel
[INFO  2026-06-24T16:55:20Z] Exiting...
[INFO  2026-06-24T16:55:20Z] Stopping config reloader...
[INFO  2026-06-24T16:55:20Z] Stopping endpoints...
Error: Output worker errored out: grpc worker errored out: gRPC error: Failed to connect to server: reconnection attempts exhausted

```

</details>

## Summary by CodeRabbit

* **New Features**
  * Added a configurable limit for gRPC reconnect retries via config/YAML, environment variable, or CLI.
  * When not set, a default retry limit is used.
* **Bug Fixes**
  * Updated gRPC reconnect logic to stop retrying after the configured limit, including clearer “give up” behavior.
  * Improved shutdown and background-task failure propagation for more reliable termination and error reporting.
* **Tests**
  * Expanded parsing, validation, environment-variable, and reconnection retry coverage for the new setting.

### CODE REVIEW

File: fact/src/config/tests.rs [hunk 1]
```diff
@@ -1756,6 +1842,22 @@ fn env_vars() {
...
+                grpc: GrpcConfig {
+                    backoff: BackoffConfig {
+                        retries_max: Some(5),
+                        ..Default::default()
+                    },
+                    ..Default::default()
+                },
+                ..Default::default()
+            },
+        ),
```
Comment (JoukoVirtanen) [inline fact/src/config/tests.rs:1868]: Add some tests with negative retries.

File: fact/src/config/mod.rs [hunk 2]
```diff
@@ -405,6 +413,12 @@ impl TryFrom<&yaml::Hash> for BackoffConfig {
                     };
                     backoff.multiplier = Some(multiplier);
                 }
+                "retries" => {
+                    let Some(retries) = v.as_i64() else {
+                        bail!("invalid grpc.backoff.retries: {v:?}");
+                    };
+                    backoff.retries_max = Some(retries as u64);
```
[the remark this answers is not in the document]
Comment (Molter73) [reply fact/src/config/mod.rs:420]: Good catch, I've made the change and added a unit test for it.

File: fact/src/config/tests.rs (hunk 1 above)
Comment (Molter73) [reply fact/src/config/tests.rs:1868]: Done

File: fact/src/output/grpc.rs [hunk 3]
```diff
@@ -30,36 +31,63 @@ struct Backoff {
...
-        let delay = self.current.min(self.max);
+    fn next(&mut self) -> Option<Duration> {
+        if self.retries_max != 0 {
+            if self.retries_curr >= self.retries_max {
+                return None;
+            }
+            self.retries_curr += 1;
+        }
+
+        let delay = self.current;
```
Comment (Stringy) [inline fact/src/output/grpc.rs]: I think this should still be `.min(self.max)`

Comment (Molter73) [reply fact/src/output/grpc.rs]: Good catch!

**Review by Stringy — APPROVED** [review]
LGTM!
